0014-fungible-token-standard.md

```diff
@@ -0,0 +1,98 @@
+
+The token owner contract can change balances of accounts with the custom token, so it can mint, burn, and move tokens. When an `AccountUpdate` that has not been created in a method of the owner contract tries to modify the balance of an account with the custom token, it needs to be approved by the owner contract. That way, the owner contract can enforce rules that all transactions with the token must satisfy (conservation of tokens being a very common example). The [`TokenContract` class](https://github.com/o1-labs/o1js/pull/1384) defines methods for token transfer, as well as for approving a whole forest of account updates. The API that we define mirrors those methods, and the reference implementation is built using the `TokenContract`.
+
+The API consists of the following `TypeScript` `interface`s:
+
+### Transferable
+This interface consists of a single function, `transfer`, which sends a specified amount of tokens from one account to another.
+
+```TypeScript
+
+interface Transferable {
+  transfer(from: PublicKey | AccountUpdate,
+           to: PublicKey | AccountUpdate,
+           amount: UInt64 | number | bigint): void;
+}
+```
+
+### Approvable
+Transactions that are either more complicated than a transfer from one account to another (such as sending tokens from one account to many accounts) or that are constructed from third-party contracts can be done using the `Approvable` interface. The workflow is to construct the individual `AccountUpdate` values, and then have them be approved by the owner contract. The interface defines three functions to cover the cases of individual `AccountUpdate`s or `AccountUpdateTree`s, an array of either of these types, or a whole `AccountUpdateForest`.
+
+```TypeScript
+interface Approvable {
+
+  approveAccountUpdate(accountUpdate: AccountUpdate | AccountUpdateTree): void;
+  approveAccountUpdates(accountUpdates: (AccountUpdate | AccountUpdateTree)[]): void;
+  approveBase(forest: AccountUpdateForest): void;
+}
+```
+
+### Administrative Interfaces
+For administrative actions, such as changing the supply or updating the `VerificationKey`, we define the `Mintable`, `Burnable`, and `Upgradeable` interfaces:
+
+```TypeScript
+interface Mintable {
+  totalSupply: State<UInt64>;
+  circulatingSupply: State<UInt64>;
+  mint: (to: PublicKey, amount: UInt64) => AccountUpdate;
+  setTotalSupply: (amount: UInt64) => void;
+}
+
+interface Burnable {
+  burn: (from: PublicKey, amount: UInt64) => AccountUpdate;
+}
+
+interface Upgradable {
+  setVerificationKey: (verificationKey: VerificationKey) => void;
```

Comment:
Probably worth keeping the recent verification key upgrade changes in mind, if you havent already https://github.com/MinaProtocol/mina/blob/develop/rfcs/0052-verification-key-permissions.md#detailed-design

Reply:
Thanks for the pointer, I'll make sure it's compatible with that rfc.

Reply:
@Trivo25 If I understand correctly, then upgrading the verification key is not really something that is special for token contracts, it's more general. Would you agree it's best to remove it from this standard?